README.md
# pomodoro_flutter
Pomodoro timer written in flutter.

## Getting Started

https://en.wikipedia.org/wiki/Pomodoro_Technique

circular_countdown_timer-0.0.9-hijacking/CHANGELOG.md
## [0.0.9] - Jan 24 2021.

* Added Optional Parameter:
    * bool named **autoStart** to handle timer start.
    * String named **textFormat** to format Countdown Text.
    * VoidCallback named **onStart** which will execute when the Countdown Starts.

## [0.0.8] - Dec 05 2020.

* Fixed **key** issue.
* Added **getTime** method in CountDownController to get the **Current Time** of Countdown Timer.
* Added Optional Parameter StrokeCap named **strokeCap** for countdown begin and end contours.

## [0.0.7] - Sep 26 2020.

* Added Optional Parameter:
    * Key named **key** for Countdown Timer.
    * bool named **isReverseAnimation** to handle Animation Direction i.e Forward or Reverse.

* Added optional int parameter named **duration** in **restart** method of CountDownController, to restart the countdown with new duration.

## [0.0.6] - Sep 06 2020.

* Added optional Color parameter **backgroundColor** to change Background Color of Countdown Widget.
* Added optional CountDownController parameter **controller** to control (i.e Pause, Resume, Restart) the Countdown Timer.

## [0.0.5] - Aug 15 2020.

* Contribution by [Joshua de Guzman](https://github.com/joshuadeguzman)
    * Added optional bool parameter **isTimerTextShown** to hide and show Timer Text.
    * Refactor Logic for Animation.

## [0.0.4] - Jun 15 2020.

* Added support to show **Hours** in Countdown Text.

## [0.0.3] - Jun 13 2020.

* Changed Parameter name from:
    * **countdownTextStyle** to **textStyle**.
    * **reverseOrder** to **isReverse**.
    * **onCountDownComplete** to **onComplete**.

## [0.0.2] - Jun 08 2020.

* Fixed Crash on **onCountDownComplete**.

## [0.0.1] - Apr 28 2020.

* Initial release.


lib/main.dart
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';


import 'pages/home.dart'; 

SharedPreferences prefs;

void main() async{
  WidgetsFlutterBinding.ensureInitialized();
  prefs = await SharedPreferences.getInstance();
  runApp(MyApp());
  if(prefs.getInt('pomodoro') == null){
    await prefs.setInt('pomodoro', 10);
  }
  if(prefs.getInt('short_break') == null){
    await prefs.setInt('short_break', 5);
  }
  if(prefs.getInt('long_break') == null){
    await prefs.setInt('long_break', 5);
  }
  if(prefs.getInt('until_long_break') == null){
    await prefs.setInt('until_long_break', 4);
  }
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Pomodoro Flutter',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Pomodoro'),
    );
  }
}



lib/pages/home.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:audioplayers/audio_cache.dart';
import 'package:circular_countdown_timer/circular_countdown_timer.dart';

import '../main.dart';
import './edit_page.dart';

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  static AudioCache player = AudioCache();
  static const alarmAudioPath = "piece-of-cake.mp3";

  Timer _timer;

  int pomodoro = prefs.getInt('pomodoro');
  int shortBreak = prefs.getInt('short_break');
  int longBreak = prefs.getInt('long_break');
  int untilLongBreak = prefs
      .getInt('until_long_break'); //when this is zero, longBreak is played.

  /*
  timerState=1=pomodoro
  timerState=0=shortBreak
  timerState=-1=longBreak
  */
  int timerState = 1;

  bool _isPlayDisabled = false; //true when the timer is running.
  bool _isPaused = false; //true when the timer is paused.
  Icon _startIcon = Icon(
      Icons.play_arrow); //this is changed depending on the state of the timer.

  CountDownController _controller = CountDownController();
  CircularCountDownTimer _animatedTimer;

  _restart() {
    _controller.restart(duration: _stateDuration());
    _controller.pause();
    _isPaused = true;
    setState(() {
      _startIcon = Icon(Icons.play_arrow);
    });
  }

  //returns the duration of state for a given timerState.
  _stateDuration() {
    if (timerState == 1) {
      return pomodoro;
    } else if (timerState == 0) {
      return shortBreak;
    } else if (timerState == -1) {
      return longBreak;
    }
  }

  //this is BAD code or is it?
  //returns the name of the state for a given timerState.
  _stateName() {
    if (timerState == 1) {
      return 'pomodoro';
    } else if (timerState == 0) {
      return 'short break';
    } else {
      return 'long break';
    }
  }

  _buildTimerUI(int key) {
    return CircularCountDownTimer(
      key: ValueKey(key),
      // Countdown duration in Seconds.
      duration: _stateDuration(),

      // Controls (i.e Start, Pause, Resume, Restart) the Countdown Timer.
      controller: _controller,

      // Width of the Countdown Widget.
      width: MediaQuery.of(context).size.width / 2,

      // Height of the Countdown Widget.
      height: MediaQuery.of(context).size.height / 2,

      // Ring Color for Countdown Widget.
      color: Colors.grey[300],

      // Filling Color for Countdown Widget.
      fillColor: Colors.blueGrey,

      // Background Color for Countdown Widget.
      backgroundColor: Colors.blueGrey[900],

      // Border Thickness of the Countdown Ring.
      strokeWidth: 20.0,

      // Begin and end contours with a flat edge and no extension.
      strokeCap: StrokeCap.round,

      // Text Style for Countdown Text.
      textStyle: TextStyle(
          fontSize: 36.0, color: Colors.white, fontWeight: FontWeight.bold),

      // Format for the Countdown Text.
      textFormat: CountdownTextFormat.MM_SS,

      // Handles Countdown Timer (true for Reverse Countdown (max to 0), false for Forward Countdown (0 to max)).
      isReverse: true,

      // Handles Animation Direction (true for Reverse Animation, false for Forward Animation).
      isReverseAnimation: true,

      // Handles visibility of the Countdown Text.
      isTimerTextShown: true,

      // Handles the timer start.
      autoStart: false,

      // This Callback will execute when the Countdown Starts.
      onStart: () {
        // Here, do whatever you want
      },

      // This Callback will execute when the Countdown Ends.
      //next timer is choosen and changed here
      onComplete: () {
        player.play(alarmAudioPath);
        setState(() {
          if (untilLongBreak > 1) {
            if (timerState == 1) {
              timerState = 0;
            } else {
              timerState = 1;
              untilLongBreak--;
            }
          } else {
            timerState = -1;
            untilLongBreak = prefs.getInt('until_long_break').toInt() + 1;
          }
          _animatedTimer = _buildTimerUI(timerState);
          _startIcon = Icon(Icons.play_arrow);
        });
        _isPlayDisabled = false;
        _isPaused = false;
      },
    );
  }

  @override
  void dispose() {
    _timer.cancel();
    super.dispose();
  }
  
  @override
  Widget build(BuildContext context) {
    _animatedTimer = _buildTimerUI(timerState);

    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
        backgroundColor: Colors.blueGrey[800],
        actions: [
          IconButton(
            icon: Icon(Icons.menu),
            onPressed: () {
              Navigator.push(context,
                      MaterialPageRoute(builder: (context) => EditPage()))
                  .then((value) {
                setState(() {
                  pomodoro = prefs.getInt('pomodoro');
                  shortBreak = prefs.getInt('short_break');
                  longBreak = prefs.getInt('long_break');
                  untilLongBreak = prefs.getInt('until_long_break');
                });
                _restart();//to update the timer with new values
              });
            },
          ),
        ],
      ),
      body: Center(
          child: GestureDetector(
        //SWIPE
        onHorizontalDragEnd: (dragEndDetails) {
          if (dragEndDetails.primaryVelocity < 0) {
            //forwards
            setState(() {
              if (timerState == 1) {
                timerState = -1;
              } else if (timerState == -1) {
                timerState = 0;
              } else if (timerState == 0) {
                timerState = 1;
              }

              _animatedTimer = _buildTimerUI(timerState);
            });
          } else if (dragEndDetails.primaryVelocity > 0) {
            //backwards
            setState(() {
              if (timerState == 1) {
                timerState = 0;
              } else if (timerState == -1) {
                timerState = 1;
              } else if (timerState == 0) {
                timerState = -1;
              }

              _animatedTimer = _buildTimerUI(timerState);
            });
          }
          _startIcon = Icon(Icons.play_arrow);
          _isPlayDisabled = false;
          _isPaused = false;
        },
        child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Text(_stateName(),
                  style: TextStyle(fontSize: 24, fontWeight: FontWeight.w900)),
              AnimatedSwitcher(
                duration: const Duration(seconds: 1),
                transitionBuilder:
                    (Widget child, Animation<double> animation) =>
                        ScaleTransition(child: child, scale: animation),
                child: _animatedTimer,
              ),
            ]),
      )),
      floatingActionButton: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Expanded(
            child: IconButton(
              icon: _startIcon,
              color: Colors.blueGrey[900],
              onPressed: () {
                //START
                if (!_isPlayDisabled && !_isPaused) {
                  _controller.start();
                  _isPlayDisabled = true;
                  setState(() {
                    _startIcon = Icon(Icons.pause);
                  });
                }
                //PAUSE
                else if (_isPlayDisabled && !_isPaused) {
                  _controller.pause();
                  _isPaused = true;
                  setState(() {
                    _startIcon = Icon(Icons.play_arrow);
                  });
                }
                //RESUME
                else if (_isPaused) {
                  _controller.resume();
                  _isPaused = false;
                  setState(() {
                    _startIcon = Icon(Icons.pause);
                  });
                }
              },
            ),
          ),
          Expanded(
            child: IconButton(
              icon: Icon(Icons.replay),
              color: Colors.blueGrey[900],
              onPressed: () => _restart(),
            ),
          ),
        ],
      ),
    );
  }
}


lib/pages/edit_page.dart
import 'package:flutter/material.dart';
import '../main.dart';

class EditPage extends StatefulWidget {
  @override
  _EditPageState createState() => _EditPageState();
}

class _EditPageState extends State<EditPage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Edit'), backgroundColor: Colors.blueGrey[800],),
      body: _editPage(),
    );
  }

  _createTimeChooser({String name, int timerLen, String prefsName}) {
    return Card(
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(15.0),
      ),
      color: Colors.blueGrey,
      child: Column(
        children: [
          Text('$name', style: TextStyle(fontSize: 16)),
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Padding(
                padding: EdgeInsets.all(20),
                child: IconButton(
                  icon: Icon(Icons.remove),
                  onPressed: () async {
                    if (timerLen > 1) {
                      setState(() {
                        timerLen--;
                      });
                      await prefs.setInt('$prefsName', timerLen);
                    }
                  },
                ),
              ),
              Padding(
                padding: EdgeInsets.all(5),
                child: Text('$timerLen', style: TextStyle(fontSize: 26)),
              ),
              Padding(
                padding: EdgeInsets.all(20),
                child: IconButton(
                  icon: Icon(Icons.add),
                  onPressed: () async {
                    setState(() {
                      timerLen++;
                    });
                    await prefs.setInt('$prefsName', timerLen);
                  },
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }

  _editPage() {
    int pomodoro = prefs.getInt('pomodoro');
    int shortBreak = prefs.getInt('short_break');
    int longBreak = prefs.getInt('long_break');
    int untilLongBreak = prefs.getInt('until_long_break');

    return Column(
      children: [
        _createTimeChooser(name: 'POMODORO \n DURATION', timerLen: pomodoro, prefsName: 'pomodoro'),
        _createTimeChooser(name: 'SHORT BREAK \n DURATION', timerLen: shortBreak, prefsName: 'short_break'),
        _createTimeChooser(name: 'LONG BREAK \n DURATION', timerLen: longBreak, prefsName: 'long_break'),
        _createTimeChooser(name: 'POMODOROS UNTIL \n LONG BREAK', timerLen: untilLongBreak, prefsName: 'until_long_break'),
      ],
    );
  }
}


circular_countdown_timer-0.0.9-hijacking/lib/custom_timer_painter.dart
import 'package:flutter/material.dart';
import 'dart:math' as math;

class CustomTimerPainter extends CustomPainter {
  CustomTimerPainter(
      {this.animation,
      this.fillColor,
      this.color,
      this.strokeWidth,
      this.strokeCap,
      this.backgroundColor})
      : super(repaint: animation);

  final Animation<double> animation;
  final Color fillColor, color, backgroundColor;
  final double strokeWidth;
  final StrokeCap strokeCap;

  @override
  void paint(Canvas canvas, Size size) {
    Paint paint = Paint()
      ..color = color
      ..strokeWidth = strokeWidth ?? 5.0
      ..strokeCap = strokeCap ?? StrokeCap.butt
      ..style = PaintingStyle.stroke;

    canvas.drawCircle(size.center(Offset.zero), size.width / 2, paint);
    double progress = (animation.value) * 2 * math.pi;
    paint.color = fillColor;
    canvas.drawArc(Offset.zero & size, math.pi * 1.5, progress, false, paint);

    if (backgroundColor != null) {
      final backgroundPaint = Paint();
      backgroundPaint.color = backgroundColor;
      canvas.drawCircle(
          size.center(Offset.zero), size.width / 2.2, backgroundPaint);
    }
  }

  @override
  bool shouldRepaint(CustomTimerPainter old) {
    return animation.value != old.animation.value ||
        color != old.color ||
        fillColor != old.fillColor;
  }
}


circular_countdown_timer-0.0.9-hijacking/lib/circular_countdown_timer.dart
library circular_countdown_timer;

import 'package:flutter/material.dart';
import 'custom_timer_painter.dart';

/// Create a Circular Countdown Timer.
class CircularCountDownTimer extends StatefulWidget {
  /// Key for Countdown Timer.
  final Key key;

  /// Filling Color for Countdown Widget.
  final Color fillColor;

  /// Ring Color for Countdown Widget.
  final Color color;

  /// Background Color for Countdown Widget.
  final Color backgroundColor;

  /// This Callback will execute when the Countdown Ends.
  final VoidCallback onComplete;

  /// This Callback will execute when the Countdown Starts.
  final VoidCallback onStart;

  /// Countdown duration in Seconds.
  final int duration;

  /// Width of the Countdown Widget.
  final double width;

  /// Height of the Countdown Widget.
  final double height;

  /// Border Thickness of the Countdown Ring.
  final double strokeWidth;

  /// Begin and end contours with a flat edge and no extension.
  final StrokeCap strokeCap;

  /// Text Style for Countdown Text.
  final TextStyle textStyle;

  /// Format for the Countdown Text.
  final String textFormat;

  /// Handles Countdown Timer (true for Reverse Countdown (max to 0), false for Forward Countdown (0 to max)).
  final bool isReverse;

  /// Handles Animation Direction (true for Reverse Animation, false for Forward Animation).
  final bool isReverseAnimation;

  /// Handles visibility of the Countdown Text.
  final bool isTimerTextShown;

  /// Controls (i.e Start, Pause, Resume, Restart) the Countdown Timer.
  final CountDownController controller;

  /// Handles the timer start.
  final bool autoStart;

  CircularCountDownTimer(
      {@required this.width,
      @required this.height,
      @required this.duration,
      @required this.fillColor,
      @required this.color,
      this.backgroundColor,
      this.isReverse = false,
      this.isReverseAnimation = false,
      this.onComplete,
      this.onStart,
      this.strokeWidth,
      this.strokeCap,
      this.textStyle,
      this.key,
      this.isTimerTextShown = true,
      this.autoStart = true,
      this.textFormat,
      this.controller})
      : assert(width != null),
        assert(height != null),
        assert(duration != null),
        assert(fillColor != null),
        assert(color != null),
        super(key: key);

  @override
  CircularCountDownTimerState createState() => CircularCountDownTimerState();
}

class CircularCountDownTimerState extends State<CircularCountDownTimer>
    with TickerProviderStateMixin {
  AnimationController _controller;
  Animation<double> _countDownAnimation;

  String get time {
    if (widget.isReverse && _controller.isDismissed) {
      if (widget.textFormat == CountdownTextFormat.MM_SS) {
        return "START";
        // return "HABEEBI";
      } else if (widget.textFormat == CountdownTextFormat.SS) {
        return "00";
      } else {
        return "00:00:00";
      }
    } else {
      Duration duration = _controller.duration * _controller.value;
      return _getTime(duration);
    }
  }

  void _setAnimation() {
    if (widget.autoStart) {
      if (widget.isReverse) {
        _controller.reverse(from: 1);
      } else {
        _controller.forward();
      }
    }
  }

  void _setAnimationDirection() {
    if ((!widget.isReverse && widget.isReverseAnimation) ||
        (widget.isReverse && !widget.isReverseAnimation)) {
      _countDownAnimation =
          Tween<double>(begin: 1, end: 0).animate(_controller);
    }
  }

  void _setController() {
    widget.controller?._state = this;
    widget.controller?._isReverse = widget.isReverse;
  }

  String _getTime(Duration duration) {
    // For HH:mm:ss format
    if (widget.textFormat == CountdownTextFormat.HH_MM_SS) {
      return '${duration.inHours.toString().padLeft(2, '0')}:${(duration.inMinutes % 60).toString().padLeft(2, '0')}:${(duration.inSeconds % 60).toString().padLeft(2, '0')}';
    }
    // For mm:ss format
    else if (widget.textFormat == CountdownTextFormat.MM_SS) {
      return '${(duration.inMinutes % 60).toString().padLeft(2, '0')}:${(duration.inSeconds % 60).toString().padLeft(2, '0')}';
    }
    // For ss format
    else if (widget.textFormat == CountdownTextFormat.SS) {
      return '${(duration.inSeconds)}';
    } else {
      // Default format
      return _defaultFormat(duration);
    }
  }

  _defaultFormat(Duration duration) {
    if (duration.inHours != 0) {
      return '${duration.inHours.toString().padLeft(2, '0')}:${(duration.inMinutes % 60).toString().padLeft(2, '0')}:${(duration.inSeconds % 60).toString().padLeft(2, '0')}';
    } else if (duration.inMinutes != 0) {
      return '${(duration.inMinutes % 60).toString().padLeft(2, '0')}:${(duration.inSeconds % 60).toString().padLeft(2, '0')}';
    } else {
      return '${duration.inSeconds % 60}';
    }
  }

  void _onStart() {
    if (widget.onStart != null) widget.onStart();
  }

  void _onComplete() {
    if (widget.onComplete != null) widget.onComplete();
  }

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      vsync: this,
      duration: Duration(seconds: widget.duration),
    );

    _controller.addStatusListener((status) {
      switch (status) {
        case AnimationStatus.forward:
          _onStart();
          break;

        case AnimationStatus.reverse:
          _onStart();
          break;

        case AnimationStatus.dismissed:
          _onComplete();
          break;
        case AnimationStatus.completed:

          /// [AnimationController]'s value is manually set to [1.0] that's why [AnimationStatus.completed] is invoked here this animation is [isReverse]
          /// Only call the [_onComplete] block when the animation is not reversed.
          if (!widget.isReverse) _onComplete();
          break;
        default:
        // Do nothing
      }
    });

    _setAnimation();
    _setAnimationDirection();
    _setController();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: widget.width,
      height: widget.height,
      child: AnimatedBuilder(
          animation: _controller,
          builder: (context, child) {
            return Align(
              child: AspectRatio(
                aspectRatio: 1.0,
                child: Stack(
                  children: <Widget>[
                    Positioned.fill(
                      child: CustomPaint(
                        painter: CustomTimerPainter(
                            animation: _countDownAnimation ?? _controller,
                            fillColor: widget.fillColor,
                            color: widget.color,
                            strokeWidth: widget.strokeWidth,
                            strokeCap: widget.strokeCap,
                            backgroundColor: widget.backgroundColor),
                      ),
                    ),
                    widget.isTimerTextShown
                        ? Align(
                            alignment: FractionalOffset.center,
                            child: Text(
                              time,
                              style: widget.textStyle ??
                                  TextStyle(
                                    fontSize: 16.0,
                                    color: Colors.black,
                                  ),
                            ),
                          )
                        : Container(),
                  ],
                ),
              ),
            );
          }),
    );
  }

  @override
  void dispose() {
    _controller.stop();
    _controller.dispose();
    super.dispose();
  }
}

/// Controls (i.e Start, Pause, Resume, Restart) the Countdown Timer.
class CountDownController {
  CircularCountDownTimerState _state;
  bool _isReverse;

  /// This Method Starts the Countdown Timer
  void start() {
    if (_isReverse) {
      _state._controller?.reverse(from: 1);
    } else {
      _state._controller?.forward(from: 0);
    }
  }

  /// This Method Pauses the Countdown Timer
  void pause() {
    _state._controller?.stop(canceled: false);
  }

  /// This Method Resumes the Countdown Timer
  void resume() {
    if (_isReverse) {
      _state._controller?.reverse(from: _state._controller.value);
    } else {
      _state._controller?.forward(from: _state._controller.value);
    }
  }

  /// This Method Restarts the Countdown Timer,
  /// Here optional int parameter **duration** is the updated duration for countdown timer

  void restart({int duration}) {
    _state._controller.duration =
        Duration(seconds: duration ?? _state._controller.duration.inSeconds);
    if (_isReverse) {
      _state._controller?.reverse(from: 1);
    } else {
      _state._controller?.forward(from: 0);
    }
  }

  /// This Method returns the **Current Time** of Countdown Timer i.e
  /// Time Used in terms of **Forward Countdown** and Time Left in terms of **Reverse Countdown**

  String getTime() {
    return _state
        ._getTime(_state._controller.duration * _state._controller?.value);
  }
}

class CountdownTextFormat {
  static const String HH_MM_SS = "HH:mm:ss";
  static const String MM_SS = "mm:ss";
  static const String SS = "ss";
}


circular_countdown_timer-0.0.9-hijacking/example/example.dart
import 'package:flutter/material.dart';
import 'package:circular_countdown_timer/circular_countdown_timer.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Circular Countdown Timer Demo',
      theme: ThemeData(
        primarySwatch: Colors.purple,
      ),
      home: MyHomePage(title: 'Circular Countdown Timer'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  CountDownController _controller = CountDownController();
  int _duration = 10;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
          child: CircularCountDownTimer(
        // Countdown duration in Seconds.
        duration: _duration,

        // Controls (i.e Start, Pause, Resume, Restart) the Countdown Timer.
        controller: _controller,

        // Width of the Countdown Widget.
        width: MediaQuery.of(context).size.width / 2,

        // Height of the Countdown Widget.
        height: MediaQuery.of(context).size.height / 2,

        // Ring Color for Countdown Widget.
        color: Colors.grey[300],

        // Filling Color for Countdown Widget.
        fillColor: Colors.purpleAccent[100],

        // Background Color for Countdown Widget.
        backgroundColor: Colors.purple[500],

        // Border Thickness of the Countdown Ring.
        strokeWidth: 20.0,

        // Begin and end contours with a flat edge and no extension.
        strokeCap: StrokeCap.round,

        // Text Style for Countdown Text.
        textStyle: TextStyle(
            fontSize: 33.0, color: Colors.white, fontWeight: FontWeight.bold),

        // Format for the Countdown Text.
        textFormat: CountdownTextFormat.SS,

        // Handles Countdown Timer (true for Reverse Countdown (max to 0), false for Forward Countdown (0 to max)).
        isReverse: false,

        // Handles Animation Direction (true for Reverse Animation, false for Forward Animation).
        isReverseAnimation: false,

        // Handles visibility of the Countdown Text.
        isTimerTextShown: true,

        // Handles the timer start.
        autoStart: false,

        // This Callback will execute when the Countdown Starts.
        onStart: () {
          // Here, do whatever you want
          print('Countdown Started');
        },

        // This Callback will execute when the Countdown Ends.
        onComplete: () {
          // Here, do whatever you want
          print('Countdown Ended');
        },
      )),
      floatingActionButton: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          SizedBox(
            width: 30,
          ),
          _button(title: "Start", onPressed: () => _controller.start()),
          SizedBox(
            width: 10,
          ),
          _button(title: "Pause", onPressed: () => _controller.pause()),
          SizedBox(
            width: 10,
          ),
          _button(title: "Resume", onPressed: () => _controller.resume()),
          SizedBox(
            width: 10,
          ),
          _button(
              title: "Restart",
              onPressed: () => _controller.restart(duration: _duration))
        ],
      ),
    );
  }

  _button({String title, VoidCallback onPressed}) {
    return Expanded(
        child: RaisedButton(
      child: Text(
        title,
        style: TextStyle(color: Colors.white),
      ),
      onPressed: onPressed,
      color: Colors.purple,
    ));
  }
}
